base/linalg/generic.jl

```diff
@@ -103,21 +103,33 @@ function generic_vecnorm1(x)
     return convert(T, sum)
 end
 
+function generic_vecnorm2(n, scale, ssq, x)
+    @inbounds for i = 1:n
+        if x[i] != 0
+            absxi = norm(x[i])
+            if scale < absxi
+                ssq = 1 + ssq*(scale/absxi)^2
+                scale = absxi
+            else
+                ssq = ssq + (absxi/scale)^2
+            end
+        end
+    end
+    return scale * sqrt(ssq)
+end
+function generic_vecnorm2{T<:Number}(x::AbstractVector{T})
+    scale = zero(norm(zero(T)))
+    ssq = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
+    return generic_vecnorm2(length(x), scale, ssq, x)
+end
 function generic_vecnorm2(x)
-    maxabs = vecnormInf(x)
-    (maxabs == 0 || isinf(maxabs)) && return maxabs
-    s = start(x)
-    (v, s) = next(x, s)
-    T = typeof(maxabs)
-    scale::promote_type(Float64, T) = 1/maxabs
-    y = norm(v)*scale
-    sum::promote_type(Float64, T) = y*y
-    while !done(x, s)
-        (v, s) = next(x, s)
-        y = norm(v)*scale
-        sum += y*y
+    n = length(x)
+    if n == 0
+        throw(ArgumentError("input must either be an AbstractVector{T<:Number} or non-empty"))
     end
-    return convert(T, maxabs * sqrt(sum))
+    scale = zero(norm(x[1]))
```

Comment:
Even with inbounds,

```julia
julia> a = Complex128[1, 2, 3]
3-element Array{Complex{Float64},1}:                                              
 1.0+0.0im                                                                        
 2.0+0.0im                                                                        
 3.0+0.0im                                                                        

julia> f(x) = (@inbounds v = zero(norm(x[1])); v)
f (generic function with 1 method)

julia> @code_llvm f(a)

define double @julia_f_21134(%jl_value_t*) {
top:
  %1 = alloca %Complex.12, align 8
  %2 = bitcast %jl_value_t* %0 to %Complex.12**
  %3 = load %Complex.12** %2, align 8
  %4 = load %Complex.12* %3, align 8
  store %Complex.12 %4, %Complex.12* %1, align 8
  %5 = call double @julia_norm_21121(%Complex.12* %1)
  ret double 0.000000e+00
}
```

I don't think type inference or llvm is smart enough to figure out `norm(::Complex)` is effect free. There's even more junk emitted without `@inbounds`.